src/controllers/services.js

```diff
@@ -0,0 +1,101 @@
+import { fetchCollections } from '../services/collections.js';
+import { fetchServices } from '../services/services.js';
+
+export const getServices = async (req, res) => {
+  const { name, termsTypes } = req.query;
+
+  const collections = await fetchCollections().catch(error => {
+    res.status(error.response?.status || 500).json({ error: `${error.message} (${error.code})` });
+  });
+
+  if (!collections) {
+    return;
+  }
+
+  const results = [];
+  const failures = [];
+
+  await Promise.all(collections.map(async ({ id: collectionId, endpoint }) => {
+    const services = await fetchServices(endpoint).catch(error => {
+      failures.push({ collection: collectionId, message: error.toString() });
+    });
+
+    if (!services) {
+      return;
```

Comment:
If an error is raised, it will be caught with the `catch` on line 19, so the execution won't be interrupted and `services` will be undefined.

I could also write it like this:

```js
   try {
      const services = await fetchServices(endpoint);

      for (const service of services) {
        if (!service.terms || (requestedName && service.name !== requestedName)) {
          continue;
        }

        const serviceTermsTypes = service.terms.map(terms => terms.type);
        const filteredServiceTermsTypes = requestedTermsTypes ? serviceTermsTypes.filter(serviceTermsType => requestedTermsTypes.includes(serviceTermsType)) : serviceTermsTypes;

        if (!filteredServiceTermsTypes.length) {
          continue;
        }

        results.push({
          collection: collectionId,
          service: {
            id: service.id,
            name: service.name,
            url: `${endpoint}/service/${encodeURIComponent(service.id)}`,
            termsTypes: filteredServiceTermsTypes,
          },
        });
      }
    } catch (error) {
      failures.push({ collection: collectionId, message: error.toString() });
    }
```

But I find the intention of catching error only for the `fetch` less clear. And the behavior is not exactly the same; If I get an error in the `for` loop it will also be caught with this code.

Reply:
I understand, thank you very much for this explanation.
From my point of view, this reinforces the relevance of a functional approach. If you're not convinced, then let's keep the current implementation 🙂